src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/WPF/ApplicationXamlFileAccessor.cs

```diff
@@ -0,0 +1,190 @@
+{
+    private readonly UnconfiguredProject _project;
+    private readonly IServiceProvider _serviceProvider;
+    private readonly IVsUIService<IVsRunningDocumentTable> _runningDocumentTable;
+    private readonly IProjectThreadingService _threadingService;
+
+    private DocData? _applicationXamlDocdata;
+    private AppDotXamlDocument? _applicationDotXamlDocument;
+
+    [ImportingConstructor]
+    public ApplicationXamlFileAccessor(
+        UnconfiguredProject project,
+        [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+        IVsUIService<IVsRunningDocumentTable> runningDocumentTable,
+        IProjectThreadingService threadingService)
+    {
+        _project = project;
+        _serviceProvider = serviceProvider;
+        _runningDocumentTable = runningDocumentTable;
+        _threadingService = threadingService;
+    }
+
+    public async Task<string?> GetStartupUriAsync()
+    {
+        await _threadingService.SwitchToUIThread();
+
+        if (await TryGetApplicationXamlFileAsync(create: false) is not AppDotXamlDocument xamlDocument)
+        {
+            return null;
+        }
+
+        return xamlDocument.GetStartupUri();
+    }
+
+    public async Task SetStartupUriAsync(string startupUri)
+    {
+        await _threadingService.SwitchToUIThread();
+
+        AppDotXamlDocument? xamlDocument = await TryGetApplicationXamlFileAsync(create: true);
+
+        if (xamlDocument is not null)
+        {
+            xamlDocument.SetStartupUri(startupUri);
+            SaveDocDataIfOnlyEditor();
+        }
+    }
+
+    public async Task<string?> GetShutdownModeAsync()
+    {
+        await _threadingService.SwitchToUIThread();
+
+        AppDotXamlDocument? xamlDocument = await TryGetApplicationXamlFileAsync(create: false);
+
+        return xamlDocument?.GetShutdownMode();
+    }
+
+    public async Task SetShutdownModeAsync(string shutdownMode)
+    {
+        await _threadingService.SwitchToUIThread();
+
+        AppDotXamlDocument? xamlDocument = await TryGetApplicationXamlFileAsync(create: true);
+
+        if (xamlDocument is not null)
+        {
+            xamlDocument.SetShutdownMode(shutdownMode);
+            SaveDocDataIfOnlyEditor();
+        }
+    }
+
+    public async Task<bool> ApplicationXamlFileExistsAsync()
+    {
+        return !string.IsNullOrEmpty(await GetFilePathAsync(create: false));
+    }
+
+    private async Task<AppDotXamlDocument?> TryGetApplicationXamlFileAsync(bool create)
+    {
+        if (_applicationDotXamlDocument is null)
+        {
+            DocData? docData = await GetDocDataAsync(create);
+
+            if (docData is not null)
+            {
+                if (docData.Buffer is IVsTextLines vsTextLines)
+                {
+                    _applicationDotXamlDocument = new AppDotXamlDocument(vsTextLines);
+                }
+            }
+        }
+
+        return _applicationDotXamlDocument;
+    }
+
+    private void SaveDocDataIfOnlyEditor()
```

Comment:
This attempts to re-create some behavior in the old property pages that may not be needed. If we're the only ones with the Application.xaml file open for editing we'll save any pending changes; but it the file is held open by others (e.g., it is open in an editor) we won't.